### ISSUE TITLE
Narrowing occurred during type inference

### ISSUE THREAD

**Issue by mxr576** [squizlabs/PHP_CodeSniffer#1785]
```
> phpcbf -p --standard=PSR2 --runtime-set ignore_errors_on_exit 1 --runtime-set ignore_warnings_on_exit 1 src
PHP Warning:  Narrowing occurred during type inference. Please file a bug report on bugs.php.net in mylib/vendor/squizlabs/php_codesniffer/autoload.php on line 167
PHP Stack trace:
PHP   1. {main}() mylib/vendor/squizlabs/php_codesniffer/bin/phpcbf:0
PHP   2. PHP_CodeSniffer\Runner->runPHPCBF() mylib/vendor/squizlabs/php_codesniffer/bin/phpcbf:18
PHP   3. PHP_CodeSniffer\Runner->init() mylib/vendor/squizlabs/php_codesniffer/src/Runner.php:176
PHP   4. PHP_CodeSniffer\Ruleset->__construct() mylib/vendor/squizlabs/php_codesniffer/src/Runner.php:294
PHP   5. PHP_CodeSniffer\Ruleset->populateTokenListeners() mylib/vendor/squizlabs/php_codesniffer/src/Ruleset.php:218
PHP   6. PHP_CodeSniffer\Standards\Squiz\Sniffs\Functions\FunctionDeclarationSniff->register() mylib/vendor/squizlabs/php_codesniffer/src/Ruleset.php:1175
PHP   7. PHP_CodeSniffer\Standards\Squiz\Sniffs\Functions\FunctionDeclarationSniff->parse() mylib/vendor/squizlabs/php_codesniffer/src/Sniffs/AbstractPatternSniff.php:91
PHP   8. PHP_CodeSniffer\Standards\Squiz\Sniffs\Functions\FunctionDeclarationSniff->createTokenPattern() mylib/vendor/squizlabs/php_codesniffer/src/Sniffs/AbstractPatternSniff.php:826
PHP   9. spl_autoload_call() mylib/vendor/squizlabs/php_codesniffer/src/Sniffs/AbstractPatternSniff.php:918
PHP  10. PHP_CodeSniffer\Autoload::load() mylib/vendor/squizlabs/php_codesniffer/src/Sniffs/AbstractPatternSniff.php:918
PHP  11. PHP_CodeSniffer\Autoload::loadFile() mylib/vendor/squizlabs/php_codesniffer/autoload.php:134
PHP  12. include() mylib/vendor/squizlabs/php_codesniffer/autoload.php:167
PHP  13. spl_autoload_call() mylib/vendor/squizlabs/php_codesniffer/src/Tokenizers/PHP.php:14
PHP  14. PHP_CodeSniffer\Autoload::load() mylib/vendor/squizlabs/php_codesniffer/src/Tokenizers/PHP.php:14
PHP  15. PHP_CodeSniffer\Autoload::loadFile() mylib/vendor/squizlabs/php_codesniffer/autoload.php:134

Warning: Narrowing occurred during type inference. Please file a bug report on bugs.php.net in mylib/vendor/squizlabs/php_codesniffer/autoload.php on line 167
```

```sh
php -v
PHP 7.1.12 (cli) (built: Nov 29 2017 11:45:18) ( NTS )
Copyright (c) 1997-2017 The PHP Group
Zend Engine v3.1.0, Copyright (c) 1998-2017 Zend Technologies
    with Zend OPcache v7.1.12, Copyright (c) 1999-2017, by Zend Technologies
    with Xdebug v2.5.5, Copyright (c) 2002-2017, by Derick Rethans
```

Installed version: 3.2.0

Everything is fine with < 3.2.0 versions.

**Comment by gsherwood** [comment]
I'm not sure what I'm supposed to do with this issue. It says to report a bug to PHP and a bit of searching tells me it's an already reported bug with PHP 7.1's opcache. So I'd suggest searching around for a solution for your PHP version. There isn't anything I can do in PHPCS to stop this I don't think. I can't replicate it on any PHP versions I've got, so it might be specific to your setup as well.

**Comment by gmponos** [comment]
I have the same issue.

> I can't replicate it on any PHP versions I've got

If there is anyway I can help.

> Everything is fine with < 3.2.0 versions.

Same for me, switching back to version 3.1.1 fixes the issue.

**Comment by gmponos** [comment]
By the way it happens to PHP 7.2. on PHP 5.6 it works normally

**Comment by simonhamp** [comment]
FYI, I'm getting this on PHP 7.1.12 with PHPCS 3.2.2.

**Comment by gsherwood** [comment]
It would be great if someone getting this error could open a bug report at bugs.php.net, as the error says. I can't replicate the issue, so I can't create a test script to expose it.

**Comment by gsherwood** [comment]
I'm going to close this. Looking around, this issue has been reported on a number of projects and none that I found were able to change anything to resolve it. If you are getting this issue, please report the bug to PHP and see if they can figure out what it is.

**Comment by jrfnl** [comment]
FYI: from what I can gather, the issue existed in PHP 7.1.12 and PHP 7.2.0

Both PHP 7.1.13 as well as 7.2.1 have been released yesterday and according to the changelog contain a fix related to the issue.

Would be great if someone could confirm that the new PHP versions solve the issue.

Refs:
* http://php.net/archive/2018.php#id2018-01-04
* http://www.php.net/ChangeLog-7.php#7.2.1
* http://www.php.net/ChangeLog-7.php#7.1.13

**Comment by disc** [comment]
@jrfnl This issue was not fixed for me on PHP 7.1.13
`php -v`

> PHP 7.1.13 (cli) (built: Jan  3 2018 11:00:58) ( NTS )
> Copyright (c) 1997-2017 The PHP Group
> Zend Engine v3.1.0, Copyright (c) 1998-2017 Zend Technologies
>     with Zend OPcache v7.1.13, Copyright (c) 1999-2017, by Zend Technologies
>     with blackfire v1.18.0~linux-x64-non_zts71, https://blackfire.io, by SensioLabs

`vendor/bin/phpcbf -n`
```
PHP Warning:  Narrowing occurred during type inference. Please file a bug report on bugs.php.net in /vagrant/StripChat/vendor/squizlabs/php_codesniffer/autoload.php on line 167
.S..........................................................   60 / 1458 (4%)
............................................................  120 / 1458 (8%)
............................................................  180 / 1458 (12%)
............................................................  240 / 1458 (16%)
............................................................  300 / 1458 (21%)
............................................................  360 / 1458 (25%)
............................................................  420 / 1458 (29%)
............................................................  480 / 1458 (33%)
............................................................  540 / 1458 (37%)
............................................................  600 / 1458 (41%)
............................................................  660 / 1458 (45%)
............................................................  720 / 1458 (49%)
............................................................  780 / 1458 (53%)
............................................................  840 / 1458 (58%)
............................................................  900 / 1458 (62%)
............................................................  960 / 1458 (66%)
............................................................ 1020 / 1458 (70%)
............................................................ 1080 / 1458 (74%)
............................................................ 1140 / 1458 (78%)
............................................................ 1200 / 1458 (82%)
............................................................ 1260 / 1458 (86%)
............................................................ 1320 / 1458 (91%)
............................................................ 1380 / 1458 (95%)
............................................................ 1440 / 1458 (99%)
..................                                           1458 / 1458 (100%)



No fixable errors were found

Time: 16.42 secs; Memory: 64.01Mb
```

**Comment by jrfnl** [comment]
@disc In that case, you may want to add a comment to the bug report in https://bugs.php.net/bug.php?id=75608

**Comment by nikic** [comment]
Can you please try to determine which file is being autoloaded when this warning occurs? E.g. by adding a var_dump with the file name and seeing what is printed directly before the warning.

**Comment by nikic** [comment]
Most likely this is fixed by https://github.com/php/php-src/commit/8a4532319dfae83ff16b2d2bbfeed062924c3c27 in 7.2-dev, which may need a 7.1 backport.

**Comment by disc** [comment]
@nikic file path is `vendor/squizlabs/php_codesniffer/src/Tokenizers/Tokenizer.php`

**Comment by nikic** [comment]
I've landed the aforementioned commit in 7.1 and confirmed that it resolves the issue with the Tokenizer.php file.

**Comment by jrfnl** [comment]
@nikic Nikita, you are an absolute star! Thanks for taking note of this issue and solving it.

**Comment by gsherwood** [comment]
@nikic Thanks a lot for looking into this.

**Comment by mxr576** [comment]
And it is available since 7.1.4.

**Comment by disc** [comment]
Since 7.1.14 it works properly
